rendergraph: use opengl node and add shaders

m0dB opened the pull request:
This is a follow up for https://github.com/mixxxdj/mixxx/pull/14007

With this PR, the allshader waveformwidget and allshader waveformrenderer classes use rendergraph at the highest level: each waveformrenderXX class is derived from a rendergraph node, and all nodes are added to as children to a top node in the waveformwidget. Rendering is that done using the rendergraph engine.

At this point all allshader waveformrenderer classes are derived from a rendergraph::OpenGLNode, and the actual rendering code is still OpenGL, and not yet using the rendergraph render functionality. So effectively the change is that the paintGL functions are called through the rendergraph engine, traversing the nodes, rather than by iterating over them directly in the waveformwidget.

The subsequent PRs will derive each of the allshader waveformrenderer classes from rendergraph::GeometryNode (or of a rendergraph::Node containing a tree of GeometryNodes). 

This PR also adds the rendergraph shaders, which will be needed by these PRs.

Commit:
use rendergraph in allshader/waveformwidget, derive allshader waveformrenderer classes from rendergraph::openglnode

Commit:
added rendergraph shaders

Commit:
avoid rendering on images with 0 or negative width or height

Commit:
fixed double/float conversions

Commit:
play nice with precompiled headers

acolombier commented on src/rendergraph/shaders/generate_shaders_gl.pl:
What do you think of using python instead of perl? The rationnal is, most of the tools script we have are either shell (sh, ps or bat for windows) or python, and I'm thinking it could be great to reduce the proliferation of languages.

Happy to merge it as is, and provide a follow up PR to translate into Python if you would want some help on the matter?

m0dB replied:
The reason for me to do this in perl is that I can do this very quickly in perl, and in python it would take me much longer. If you feel like porting it to python, be my guest!

Holzhaus replied:
I'll send you a PR

Holzhaus replied:
https://github.com/m0dB/mixxx/pull/18

acolombier commented on src/rendergraph/shaders/generated_shaders_gl.cmake:
```diff
@@ -0,0 +1,13 @@
+  rgba.frag.gl
+  texture.frag.gl
+  unicolor.frag.gl
+)
```
How would you feel about automating the generation of the `.gl` script using something like this?

```cmake
execute_process(
  COMMAND ./generate_shaders_gl.pl
  RESULT_VARIABLE result_generate_gl
)
if (${result_generate_gl})
  message(FATAL_ERROR "Unable to generate the GL files: ${result_generate_gl}")
endif()
```

Alternatively, what do you think of auto generating this in the `pre-commit` hooks? (Note: using Python for the script might be a requirement, but this needs to be checked)

```yaml
      - id: generate-rendergraph-gl
        name: generate-rendergraph-gl
        description: "Generate GL files for rendergraph saher"
        entry: ./src/rendergraph/shaders/generate_shaders_gl.pl
        require_serial: true
        language: perl
        files: ^src/rendergraph/shaders/pattern.frag/.*\.(frag|vert)$
```

m0dB replied:
I don't know if it is worth the effort. I can simply commit the generated shaders. If the need arises to add more shaders, we can improve on this. Also note that this is only a temporary solution: With Qt >=6.6 we can use the qsb bundles from C++ directly.

acolombier replied:
As we've just updated to Ubuntu 24.04 (Qt 6.4), it may be 2 years+ before the minimum Qt version is 6.6+. I don't mind updating this in a follow up PR tho.

m0dB replied:
Thanks, yes, doing this in a follow up PR would be best.

Holzhaus replied:
pre-commit is the better option, as it will only run on changes, not every time you build. Also, I'm not sure that python is already a build-time dependency.

I included the pre-commit hook in https://github.com/m0dB/mixxx/pull/18

acolombier reviewed (commented):
Looking good, just need to make a quick manual testing.

Holzhaus committed:
chore(rendergraph): Port `generate_shaders_gl.pl` to Python

Holzhaus committed:
chore(rendergraph): Order `generated_shaders_gl.cmake` alphabetically

Holzhaus committed:
ci(pre-commit): Automatically update rendergraph shaders on changes

acolombier commented:
I'm getting a segfault at startup

```
Thread 1 "mixxx" received signal SIGSEGV, Segmentation fault.
0x00005555563430eb in allshader::WaveformWidget::WaveformWidget (this=0x55558e6fcca0, parent=<optimised out>, type=WaveformWidgetType::RGB, group=..., options=...) at /home/antoine/dev/mixxx/src/waveform/widgets/allshader/waveformwidget.cpp:52
52	    if (m_pWaveformRendererSignal && m_pWaveformRendererSignal->supportsSlip()) {
(gdb) bt
#0  0x00005555563430eb in allshader::WaveformWidget::WaveformWidget(QWidget*, WaveformWidgetType::Type, QString const&, QFlags<allshader::WaveformRendererSignalBase::Option>)
    (this=0x55558e6fcca0, parent=<optimised out>, type=WaveformWidgetType::RGB, group=<optimised out>, options=...) at /home/antoine/dev/mixxx/src/waveform/widgets/allshader/waveformwidget.cpp:52
#1  0x0000555556268a56 in WaveformWidgetFactory::createAllshaderWaveformWidget(WaveformWidgetType::Type, WWaveformViewer*) (this=this@entry=0x55555c230cc0, type=type@entry=WaveformWidgetType::RGB, viewer=viewer@entry=0x55558abbef60)
    at /home/antoine/dev/mixxx/src/waveform/waveformwidgetfactory.cpp:1005
#2  0x0000555556268eb6 in WaveformWidgetFactory::createRGBWaveformWidget(WWaveformViewer*) (this=this@entry=0x55555c230cc0, viewer=viewer@entry=0x55558abbef60) at /home/antoine/dev/mixxx/src/waveform/waveformwidgetfactory.cpp:1063
#3  0x0000555556267305 in WaveformWidgetFactory::createWaveformWidget(WaveformWidgetType::Type, WWaveformViewer*) (this=this@entry=0x55555c230cc0, type=WaveformWidgetType::RGB, viewer=viewer@entry=0x55558abbef60)
    at /home/antoine/dev/mixxx/src/waveform/waveformwidgetfactory.cpp:1141
#4  0x0000555556267a90 in WaveformWidgetFactory::setWidgetTypeFromHandle(int, bool) (this=this@entry=0x55555c230cc0, handleIndex=handleIndex@entry=0, force=<optimised out>) at /home/antoine/dev/mixxx/src/waveform/waveformwidgetfactory.cpp:617
#5  0x0000555556266df4 in WaveformWidgetFactory::setWidgetTypeFromConfig() (this=0x55555c230cc0) at /home/antoine/dev/mixxx/src/waveform/waveformwidgetfactory.cpp:583
#6  0x00007ffff17ac023 in  () at /lib/x86_64-linux-gnu/libQt6Core.so.6
#7  0x000055555603e6f4 in MixxxMainWindow::skinLoaded() (this=0x7fffffffd6b0) at mixxx-lib_autogen/include/moc_mixxxmainwindow.cpp:272
#8  MixxxMainWindow::loadConfiguredSkin() (this=this@entry=0x7fffffffd6b0) at /home/antoine/dev/mixxx/src/mixxxmainwindow.cpp:1234
#9  0x000055555603b66a in MixxxMainWindow::initialize() (this=0x7fffffffd6b0) at /home/antoine/dev/mixxx/src/mixxxmainwindow.cpp:324
#10 0x00007ffff17ac023 in  () at /lib/x86_64-linux-gnu/libQt6Core.so.6
#11 0x00007ffff4677ada in QOpenGLWindow::resizeEvent(QResizeEvent*) () at /lib/x86_64-linux-gnu/libQt6OpenGL.so.6
#12 0x00007ffff1e10858 in QWindow::event(QEvent*) () at /lib/x86_64-linux-gnu/libQt6Gui.so.6
#13 0x0000555556344557 in OpenGLWindow::event(QEvent*) (this=0x353167223d646920, pEv=0x0) at /home/antoine/dev/mixxx/src/widget/openglwindow.cpp:70
#14 0x00007ffff4f7fd36 in QApplicationPrivate::notify_helper(QObject*, QEvent*) () at /lib/x86_64-linux-gnu/libQt6Widgets.so.6
#15 0x000055555602bfcf in MixxxApplication::notify(QObject*, QEvent*) (this=0x7fffffffd630, pTarget=0x55555baf1d40, pEvent=0x7fffffffd2f0) at /home/antoine/dev/mixxx/src/mixxxapplication.cpp:201
#16 0x00007ffff1753a58 in QCoreApplication::notifyInternal2(QObject*, QEvent*) () at /lib/x86_64-linux-gnu/libQt6Core.so.6
#17 0x00007ffff1dd0300 in QGuiApplicationPrivate::processExposeEvent(QWindowSystemInterfacePrivate::ExposeEvent*) () at /lib/x86_64-linux-gnu/libQt6Gui.so.6
#18 0x00007ffff1e13c8c in QWindowSystemInterface::sendWindowSystemEvents(QFlags<QEventLoop::ProcessEventsFlag>) () at /lib/x86_64-linux-gnu/libQt6Gui.so.6
#19 0x00007fffe925c86e in  () at /usr/lib/x86_64-linux-gnu/qt6/plugins/platforms/../../../libQt6XcbQpa.so.6
#20 0x00007ffff5d1bd3b in g_main_context_dispatch () at /lib/x86_64-linux-gnu/libglib-2.0.so.0
#21 0x00007ffff5d712b8 in  () at /lib/x86_64-linux-gnu/libglib-2.0.so.0
#22 0x00007ffff5d193e3 in g_main_context_iteration () at /lib/x86_64-linux-gnu/libglib-2.0.so.0
#23 0x00007ffff197beae in QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) () at /lib/x86_64-linux-gnu/libQt6Core.so.6
#24 0x00007ffff1760adb in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) () at /lib/x86_64-linux-gnu/libQt6Core.so.6
#25 0x00007ffff175c0f3 in QCoreApplication::exec() () at /lib/x86_64-linux-gnu/libQt6Core.so.6
#26 0x0000555555b427e4 in (anonymous namespace)::runMixxx(MixxxApplication*, CmdlineArgs const&) (pApp=<optimised out>, args=<optimised out>) at /home/antoine/dev/mixxx/src/main.cpp:108
#27 main(int, char**) (argc=<optimised out>, argv=<optimised out>) at /home/antoine/dev/mixxx/src/main.cpp:243
```

m0dB commented:
Strange. I’ll look into this tomorrow.

acolombier commented:
I had a look and this patch seems to fix the crash

```diff
diff --git a/src/waveform/widgets/allshader/waveformwidget.cpp b/src/waveform/widgets/allshader/waveformwidget.cpp
index e6f5536a71..fed38fae1e 100644
--- a/src/waveform/widgets/allshader/waveformwidget.cpp
+++ b/src/waveform/widgets/allshader/waveformwidget.cpp
@@ -25,7 +25,7 @@ WaveformWidget::WaveformWidget(QWidget* parent,
         WaveformWidgetType::Type type,
         const QString& group,
         WaveformRendererSignalBase::Options options)
-        : WGLWidget(parent), WaveformWidgetAbstract(group) {
+        : WGLWidget(parent), WaveformWidgetAbstract(group), m_pWaveformRendererSignal(nullptr) {
     auto pTopNode = std::make_unique<rendergraph::Node>();
     auto pOpacityNode = std::make_unique<rendergraph::OpacityNode>();
 ```

Once applied, everything looks great!

m0dB commented:
Thanks, @acolombier , a stupid omission on my side indeed.

m0dB commented:
@Holzhaus This is failing through precommit. In fact, I am not sure if it is a good idea to use precommit for this. It seems overkill. I'll remove it from precommit.

Commit:
remove generate_shaders_gl from precommit

Commit:
moved generate_shaders_gl.py script to rendergraph/tools, updated readme

m0dB commented:
When executing manually, the script fails with:
```
Traceback (most recent call last):
  File "/Users/maarten/git/mixxx/src/rendergraph/tools/./generate_shaders_gl.py", line 23, in <module>
    executable_name: str, additional_paths: list[str] | None = None
TypeError: unsupported operand type(s) for |: 'types.GenericAlias' and 'NoneType'
```

Commit:
add missing pointer initialization

Holzhaus commented:
> When executing manually, the script fails with:
> ```
> Traceback (most recent call last):
>   File "/Users/maarten/git/mixxx/src/rendergraph/tools/./generate_shaders_gl.py", line 23, in <module>
>     executable_name: str, additional_paths: list[str] | None = None
> TypeError: unsupported operand type(s) for |: 'types.GenericAlias' and 'NoneType'
> ```
> 

Your Python version is very old apparently.

m0dB commented:
> Your Python version is very old apparently.

3.9 which is what is shipped with macOS.

Holzhaus commented:
Exactly, union type expessions are supported since 3.10 (https://docs.python.org/3/library/stdtypes.html#types-union) from October 2021. I will quickly make another PR to use the old style syntax.

Holzhaus committed:
fix(generate_shaders_gl): Replace new union expression with old syntax

This makes it compatible with Python 3.9 (default on macOS).

Holzhaus commented:
https://github.com/m0dB/mixxx/pull/19

Holzhaus commented:
> @Holzhaus This is failing through precommit. In fact, I am not sure if it is a good idea to use precommit for this. It seems overkill. I'll remove it from precommit.

Why? It ensures that the shaders are regenerated if you touch the source files. It won't run if you don't modify the files.

Holzhaus commented on src/rendergraph/tools/generate_shaders_gl.py:
```diff
@@ -0,0 +1,159 @@
+#!/usr/bin/env python3
```
Please move this to the `src/tools` directory in the root of the repo where our other tools are placed already. IMHO it don't belong in a subdirectory deep in the source tree where it's hard to find .

m0dB replied:
Rendergraph is completely self-contained (except for the include of util/assert.h), that’s why I moved it here. After all, the shaders themselves are also here. Anyway, I will move it back.

m0dB commented:
> Why? It ensures that the shaders are regenerated if you touch the source files. It won't run if you don't modify the files. 

There is no need to touch these files, or very sporadically.

But also it requires qsb and spirv to be installed in the build agents, as precommit is run there too. And this is clearly not the case.

Holzhaus commented:
> > Why? It ensures that the shaders are regenerated if you touch the source files. It won't run if you don't modify the files.
> 
> There is no need to touch these files, or very sporadically.
> 
> But also it requires qsb and spirv to be installed in the build agents, as precommit is run there too. And this is clearly not the case.

Ah, in that case we can add it to the `SKIP` list in the github action or update the docker container: https://github.com/mixxxdj/mixxx/blob/main/.github/workflows/pre-commit.yml#L18

But anyway, if you don't think it's needed we can also remove it.

Commit:
moved script back, edited info, edited readme

m0dB commented:
Moved it back and edited documentation (in python and readme). Sorry for the noise, I should have discussed this before starting moving things around. I hope it's all okay now and we can merge this soon, so I can rebased by branches that   port the allshader/waveformrenderers to rendergraph and create the corresponding MRs.

Swiftb0y commented on src/rendergraph/tools/generate_shaders_gl.py:
This is a copy of `tools/rg_generate_shaders_gl.py`?

m0dB replied:
yes, a leftover, thanks for spotting.

Swiftb0y commented on tools/rg_generate_shaders_gl.py:
^^

Swiftb0y commented on src/waveform/widgets/allshader/waveformwidget.cpp:
```diff
@@ -26,53 +25,70 @@ WaveformWidget::WaveformWidget(QWidget* parent,
 
     m_initSuccess = init();
+
+    pTopNode->appendChildNode(std::move(pOpacityNode));
+    m_pOpacityNode = static_cast<rendergraph::OpacityNode*>(pTopNode->lastChild());
```
I see this seems to be a frequent usecase, does this warrant an API change so something like this is possible?
```suggestion
    m_pOpacityNode = pTopNode->appendChildNode(std::move(pOpacityNode));
```

That would be fairly straightforward with a template like this:
```c++
template <typename T>
T* Node::appendChildNode(std::unique_ptr<T> ptr) {
    // call the non-template type-erased code here
    // ... and then return the leaked pointer in case the caller needs it
    return ptr.get();
}
```
This would avoid issues where the two lines might get detached from each other (or `lastChild` just doesn't return the right object for some reason) and the `static_cast` goes wrong. This API seems much safer than the current solution without much extra complexity. Wdyt?

m0dB replied:
Very nice and elegant! I will make the change!

m0dB replied:
done

Swiftb0y commented on src/waveform/widgets/allshader/waveformwidget.cpp:
```diff
@@ -26,53 +25,70 @@ WaveformWidget::WaveformWidget(QWidget* parent,
-        delete pRenderer;
-    }
     m_rendererStack.clear();
+    m_pEngine.reset();
```
is the explicit `.reset` needed for correctness/safety? If so please add a comment explaining why. If not, then the reset should be safe to remove.

m0dB replied:
I don't remember, I will give it a try.

m0dB replied:
the destruction needs to happen inside the OpenGL context.

Swiftb0y replied:
and the context is gone after the end of the user-defined constructor? Or what exactly is the problem?

m0dB replied:
The context is gone after the call to `doneCurrent()` (the last line of the destructor).

Swiftb0y replied:
I see. Well then a comment would be justified why the special ordering is needed.

Swiftb0y commented on src/waveform/widgets/allshader/waveformwidget.cpp:
```diff
@@ -117,31 +135,30 @@ mixxx::Duration WaveformWidget::render() {
-    }
+}
+
+void WaveformWidget::resizeRenderer(int, int, float) {
+    // defer to resizeGL
 }
```
what do you mean `defer`? Does this assume `resizeGL` is called anyways shortly after?

m0dB replied:
I don't remember. I will investigate and improve the comment.

m0dB replied:
yeah, this is called when the widget is resized, but in the case of wglwidgets we also get the resizeGL call, which is more useful as it has the right context. I'll clarify in the comment.

Swiftb0y commented on src/waveform/widgets/allshader/waveformwidget.cpp:
```diff
@@ -117,31 +135,30 @@ mixxx::Duration WaveformWidget::render() {
-    for (auto* pRenderer : std::as_const(m_rendererStack)) {
-        pRenderer->allshaderWaveformRenderer()->resizeGL(w, h);
-    }
+    w = static_cast<int>(std::lround(static_cast<qreal>(w) / devicePixelRatioF()));
+    h = static_cast<int>(std::lround(static_cast<qreal>(h) / devicePixelRatioF()));
```
`makeRoundToPixel`?

m0dB replied:
Nope, not the same.

Swiftb0y commented on src/waveform/widgets/allshader/waveformwidget.h:
```diff
@@ -42,12 +48,31 @@ class allshader::WaveformWidget final : public ::WGLWidget,
+    template<class T_Renderer, typename... Args>
+    inline std::unique_ptr<T_Renderer> addWaveformSignalRendererNode(Args&&... args) {
+        auto pRenderer = addRenderer<T_Renderer>(std::forward<Args>(args)...);
+        if (!m_pWaveformRendererSignal) {
+            m_pWaveformRendererSignal = pRenderer;
+        }
```
can you explain why this is only added if with this condition?

m0dB replied:
This template function is called twice, first for the normal signal renderer, and later for the slip signal renderer.

But this is convoluted and not clear. I will try to improve this.

m0dB replied:
this was a bit tricky, but I think I found a relatively elegant solution.

Swiftb0y commented on src/waveform/renderers/allshader/waveformrendermark.cpp:
```diff
@@ -398,9 +403,14 @@ void allshader::WaveformRenderMark::updatePlayPosMarkTexture() {
-    QImage image(static_cast<int>(imgwidth * devicePixelRatio),
-            static_cast<int>(imgheight * devicePixelRatio),
-            QImage::Format_ARGB32_Premultiplied);
+    const QSize size{static_cast<int>(std::lround(imgwidth * devicePixelRatio)),
+            static_cast<int>(std::lround(imgheight * devicePixelRatio))};
```
Also here `makeRoundToPixel`? (same in the other renders)

m0dB replied:
No, not the same.

m0dB replied:
(Note that roundToPixel does both the division and a multiplication with the devicePixelRatio)

Swiftb0y replied:
ah right. Sorry.

Commit:
removed all Q_UNUSED

m0dB commented on src/waveform/renderers/allshader/waveformrenderersignalbase.cpp:
```diff
@@ -1,8 +1,16 @@
         : ::WaveformRendererSignalBase(waveformWidget) {
 }
+
+void WaveformRendererSignalBase::draw(QPainter* painter, QPaintEvent* event) {
+    Q_UNUSED(painter);
+    Q_UNUSED(event);
+    DEBUG_ASSERT(false);
+}
```
done

Commit:
removed copy of generate_shaders_gl.py

Commit:
added appendChildNode with templatized result as suggested by Nico, cleared up code when creating a waveformsignalrenderer, and appending it as a node

Commit:
comment

Commit:
comment

Swiftb0y commented on src/waveform/widgets/allshader/waveformwidget.cpp:
```diff
@@ -21,6 +21,11 @@
 
 namespace allshader {
 
+std::unique_ptr<rendergraph::BaseNode> convert(
+        std::unique_ptr<WaveformRendererSignalBase>&& pRenderer) {
+    return std::unique_ptr<rendergraph::BaseNode>(pRenderer.release()->asNode());
+}
```
I'm split on this solution. We can either do this, which involves a virtual function call, or we can do a `qobject_cast` or `dynamic_cast`. Do you have an estimate on how it compares? Is it possible fix the object hierarchy instead so we don't have to do the horizontal cast?

m0dB replied:
The object hierarchy is horrible, and I would love to improve it, but it is not obvious and definitely outside the scope of this PR.

Note that this is only called during construction of the widget, so it really shouldn't be an issue.

Swiftb0y replied:
Fair, though I'd still like to know your opinion on why you prefer this (along with its boilerplate) over a `dynamic_cast`.

m0dB replied:
Revisiting this, I have no idea why I did it like this 😅 and indeed a simple dynamic_cast works just as well.

m0dB commented:
I think I have addressed all issues, either with code changes, added comments, or explanations. I would really like to see this merged so I create the PRs that migrate all waveformrenderers to rendergraph.

Swiftb0y commented:
Yup I agree. I resolved everything addressed and added responses to the still open topics.

Commit:
fix post rebase

Commit:
remove needless convert and asNode

m0dB commented:
There are no unresolved issues left and this should be ready to be merged. But main did change quite a bit in the meantime and there were conflicts, so I have rebased this on main. I know it's not ideal for the review history but I hope it's okay in this case.

After rebasing everything still seems to work as expected, but it might be good if you (@acolombier ?) can give it a smoke test before merging.

Holzhaus reviewed (changes requested):
Unfortunately this breaks the build for me. I'm using CMake 3.31.4 and Qt 6.8.1.

<details><summary>CMake log</summary>

```bash
$ cmake -DQT6=ON -DWARNINGS_FATAL=ON -DDEBUG_ASSERTIONS_FATAL=ON .. && cmake --build . -j $(nproc) && ctest -j $(nproc)
-- CMAKE_VERSION: 3.31.4
-- Setting CMAKE_BUILD_TYPE to 'Debug' as none was specified.
-- The C compiler identification is GNU 14.2.1
-- The CXX compiler identification is GNU 14.2.1
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Optimization level: portable
-- Enabling SSE2 CPU optimizations (>= Pentium 4)
-- Found ccache: /usr/bin/ccache
-- Support for ccache: ON
-- Selecting mold as linker
-- Enabling QML Debugging! This poses a security risk as Mixxx will open a TCP port for debugging
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE
-- Performing Test HAVE_STDATOMIC
-- Performing Test HAVE_STDATOMIC - Success
-- Found WrapAtomic: TRUE
-- Found Chromaprint: /usr/lib/libchromaprint.so (found version "1.5.1")
-- Found SQLite3: /usr/include (found version "3.48.0")
-- Could NOT find DjInterop (missing: DjInterop_DIR)
-- Could NOT find DjInterop (missing: DjInterop_LIBRARY DjInterop_INCLUDE_DIR DjInterop_VERSION) (Required is exact version "0.22.1")
-- Building libdjinterop sources (with system SQLite) fetched from GitHub
-- Found ZLIB: /usr/lib/libz.so (found suitable version "1.3.1", minimum required is "1.2.8")
-- Found Ebur128: /usr/lib/libebur128.so (found version "1.2.6")
-- Found KeyFinder: /usr/lib/libkeyfinder.so (Required is at least version "2.2.4")
-- Found FLAC: /usr/lib/libFLAC.so (found version "1.4.3")
-- Found mp3lame: /usr/lib/libmp3lame.so
-- Found OpenGL: /usr/lib/libOpenGL.so
-- Found Ogg: /usr/lib/libogg.so (found version "1.3.5")
-- Found Vorbis: /usr/include  found components: vorbis vorbisenc vorbisfile
-- Found PortAudio: /usr/lib/libportaudio.so (found version "19")
-- Found PortMidi: /usr/lib/libportmidi.so
-- Found WrapOpenGL: TRUE
-- Found XKB: /usr/lib/libxkbcommon.so (found suitable version "1.7.0", minimum required is "0.5.0")
-- Found WrapVulkanHeaders: /usr/include
-- Found Cups: /usr/lib/libcups.so (found version "2.4.11")
rgbsignal_qml.frag -> rgbsignal_qml.frag.qsb exposed as :/shaders/rgbsignal_qml.frag.qsb
CMake Warning (dev) at /usr/lib/cmake/Qt6Core/Qt6CoreMacros.cmake:3353 (message):
  Qt policy QTP0004 is not set: You need qmldir files for each extra
  directory that contains .qml files for your module.  Check
  https://doc.qt.io/qt-6/qt-cmake-policy-qtp0004.html for policy details.
  Use the qt_policy command to set the policy and suppress this warning.

Call Stack (most recent call first):
  /usr/lib/cmake/Qt6Qml/Qt6QmlMacros.cmake:3299 (__qt_internal_setup_policy)
  /usr/lib/cmake/Qt6Qml/Qt6QmlMacros.cmake:896 (qt6_target_qml_sources)
  /usr/lib/cmake/Qt6Qml/Qt6QmlMacros.cmake:1232 (qt6_add_qml_module)
  CMakeLists.txt:3205 (qt_add_qml_module)
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at /usr/lib/cmake/Qt6Core/Qt6CoreMacros.cmake:3353 (message):
  Qt policy QTP0004 is not set: You need qmldir files for each extra
  directory that contains .qml files for your module.  Check
  https://doc.qt.io/qt-6/qt-cmake-policy-qtp0004.html for policy details.
  Use the qt_policy command to set the policy and suppress this warning.

Call Stack (most recent call first):
  /usr/lib/cmake/Qt6Qml/Qt6QmlMacros.cmake:3299 (__qt_internal_setup_policy)
  /usr/lib/cmake/Qt6Qml/Qt6QmlMacros.cmake:896 (qt6_target_qml_sources)
  /usr/lib/cmake/Qt6Qml/Qt6QmlMacros.cmake:1232 (qt6_add_qml_module)
  CMakeLists.txt:3239 (qt_add_qml_module)
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Found X11: /usr/include
-- Looking for XOpenDisplay in /usr/lib/libX11.so;/usr/lib/libXext.so
-- Looking for XOpenDisplay in /usr/lib/libX11.so;/usr/lib/libXext.so - found
-- Looking for gethostbyname
-- Looking for gethostbyname - found
-- Looking for connect
-- Looking for connect - found
-- Looking for remove
-- Looking for remove - found
-- Looking for shmat
-- Looking for shmat - found
-- Looking for IceConnectionNumber in ICE
-- Looking for IceConnectionNumber in ICE - found
-- Found rubberband: /usr/lib/librubberband.so (found version "4.0.0")
-- Found SndFile: /usr/lib/libsndfile.so (found version "1.2.2")
-- Found SoundTouch: /usr/lib/libSoundTouch.so (found suitable version "2.3.3", minimum required is "2.1.2")
-- Found TagLib: /usr/lib/taglib1/libtag.so (found suitable version "1.13.1", minimum required is "1.11")
-- Found Upower: /usr/lib/libupower-glib.so (found version "1.90.7")
-- Found GLIB: /usr/include/glib-2.0;/usr/lib/glib-2.0/include (found version "2.82.4")
-- Could NOT find MP4 (missing: MP4_LIBRARY MP4_INCLUDE_DIR)
-- Found MP4v2: /usr/lib/libmp4v2.so (found version "2.1.3")
-- Searching for FFMPEG components
--   - libavcodec 61.19.100 found.
--   - libavformat 61.7.100 found.
--   - libavdevice 61.3.100 found.
--   - libavutil 59.39.100 found.
--   - libavfilter 10.4.100 found.
--   - libswscale 8.3.100 found.
--   - libswresample 5.3.100 found.
-- Found FFMPEG: /usr/lib/libavcodec.so;/usr/lib/libavformat.so;/usr/lib/libavutil.so;/usr/lib/libswresample.so
-- Found lilv: /usr/lib/liblilv-0.so
-- Could NOT find Shoutidjc (missing: Shoutidjc_LIBRARY Shoutidjc_INCLUDE_DIR)
-- Using internal libshout-idjc
-- Found OpenSSL: /usr/lib/libcrypto.so (found version "3.4.0")
-- Looking for __GNU_LIBRARY__
-- Looking for __GNU_LIBRARY__ - found
-- Looking for arpa/inet.h
-- Looking for arpa/inet.h - found
-- Looking for inttypes.h
-- Looking for inttypes.h - found
-- Looking for stdint.h
-- Looking for stdint.h - found
-- Looking for sys/select.h
-- Looking for sys/select.h - found
-- Looking for sys/socket.h
-- Looking for sys/socket.h - found
-- Looking for winsock2.h
-- Looking for winsock2.h - not found
-- Looking for sys/uio.h
-- Looking for sys/uio.h - found
-- Looking for writev
-- Looking for writev - found
-- Looking for sys/time.h
-- Looking for sys/time.h - found
-- Looking for gettimeofday
-- Looking for gettimeofday - found
-- Looking for sys/timeb.h
-- Looking for sys/timeb.h - found
-- Looking for ftime
-- Looking for ftime - found
-- Looking for time.h
-- Looking for time.h - found
-- Looking for pthread_spin_init
-- Looking for pthread_spin_init - not found
-- Looking for unistd.h
-- Looking for unistd.h - found
-- Looking for poll
-- Looking for poll - not found
-- Looking for endhostent
-- Looking for endhostent - found
-- Looking for getaddrinfo
-- Looking for getaddrinfo - found
-- Looking for getnameinfo
-- Looking for getnameinfo - found
-- Looking for inet_aton
-- Looking for inet_aton - found
-- Looking for inet_pton
-- Looking for inet_pton - found
-- Looking for sethostent
-- Looking for sethostent - found
-- Looking for strcasestr
-- Looking for strcasestr - found
-- Looking for X509_check_host
-- Looking for X509_check_host - not found
-- Found OpusFile: /usr/lib/libopusfile.so (found version "0.12")
-- Found Opus: /usr/lib/libopus.so (found version "1.5.2")
-- Found MAD: /usr/lib/libmad.so (found version "0.15.1b")
-- Found ID3Tag: /usr/lib/libid3tag.so (found version "0.16.3")
-- Found Modplug: /usr/lib/libmodplug.so (found version "0.8.9.0")
-- Found LibUSB: /usr/lib/libusb-1.0.so (found version "1.0.27")
-- Found hidapi: /usr/lib/libhidapi-libusb.so (found suitable version "0.14.0", minimum required is "0.14.0")
-- Found Libudev: /usr/lib/libudev.so
-- Qt version 6.8
-- Using QShader to load qsb shaders for opengl
CMake Error at /usr/lib/cmake/Qt6Core/Qt6CoreMacros.cmake:2127 (get_target_property):
  get_target_property() called with non-existent target "rendergraph_sg".
Call Stack (most recent call first):
  /usr/lib/cmake/Qt6ShaderToolsTools/Qt6ShaderToolsMacros.cmake:103 (_qt_internal_expose_source_file_to_ide)
  /usr/lib/cmake/Qt6ShaderToolsTools/Qt6ShaderToolsMacros.cmake:354 (_qt_internal_add_shaders_impl)
  src/rendergraph/shaders/CMakeLists.txt:17 (qt6_add_shaders)


pattern.frag -> pattern.frag.qsb exposed as :/shaders/rendergraph/pattern.frag.qsb
CMake Error at /usr/lib/cmake/Qt6Core/Qt6CoreMacros.cmake:2127 (get_target_property):
  get_target_property() called with non-existent target "rendergraph_sg".
Call Stack (most recent call first):
  /usr/lib/cmake/Qt6ShaderToolsTools/Qt6ShaderToolsMacros.cmake:103 (_qt_internal_expose_source_file_to_ide)
  /usr/lib/cmake/Qt6ShaderToolsTools/Qt6ShaderToolsMacros.cmake:354 (_qt_internal_add_shaders_impl)
  src/rendergraph/shaders/CMakeLists.txt:17 (qt6_add_shaders)


pattern.vert -> pattern.vert.qsb exposed as :/shaders/rendergraph/pattern.vert.qsb
CMake Error at /usr/lib/cmake/Qt6Core/Qt6CoreMacros.cmake:2127 (get_target_property):
  get_target_property() called with non-existent target "rendergraph_sg".
Call Stack (most recent call first):
  /usr/lib/cmake/Qt6ShaderToolsTools/Qt6ShaderToolsMacros.cmake:103 (_qt_internal_expose_source_file_to_ide)
  /usr/lib/cmake/Qt6ShaderToolsTools/Qt6ShaderToolsMacros.cmake:354 (_qt_internal_add_shaders_impl)
  src/rendergraph/shaders/CMakeLists.txt:17 (qt6_add_shaders)


rgb.frag -> rgb.frag.qsb exposed as :/shaders/rendergraph/rgb.frag.qsb
CMake Error at /usr/lib/cmake/Qt6Core/Qt6CoreMacros.cmake:2127 (get_target_property):
  get_target_property() called with non-existent target "rendergraph_sg".
Call Stack (most recent call first):
  /usr/lib/cmake/Qt6ShaderToolsTools/Qt6ShaderToolsMacros.cmake:103 (_qt_internal_expose_source_file_to_ide)
  /usr/lib/cmake/Qt6ShaderToolsTools/Qt6ShaderToolsMacros.cmake:354 (_qt_internal_add_shaders_impl)
  src/rendergraph/shaders/CMakeLists.txt:17 (qt6_add_shaders)


rgb.vert -> rgb.vert.qsb exposed as :/shaders/rendergraph/rgb.vert.qsb
CMake Error at /usr/lib/cmake/Qt6Core/Qt6CoreMacros.cmake:2127 (get_target_property):
  get_target_property() called with non-existent target "rendergraph_sg".
Call Stack (most recent call first):
  /usr/lib/cmake/Qt6ShaderToolsTools/Qt6ShaderToolsMacros.cmake:103 (_qt_internal_expose_source_file_to_ide)
  /usr/lib/cmake/Qt6ShaderToolsTools/Qt6ShaderToolsMacros.cmake:354 (_qt_internal_add_shaders_impl)
  src/rendergraph/shaders/CMakeLists.txt:17 (qt6_add_shaders)


rgba.frag -> rgba.frag.qsb exposed as :/shaders/rendergraph/rgba.frag.qsb
CMake Error at /usr/lib/cmake/Qt6Core/Qt6CoreMacros.cmake:2127 (get_target_property):
  get_target_property() called with non-existent target "rendergraph_sg".
Call Stack (most recent call first):
  /usr/lib/cmake/Qt6ShaderToolsTools/Qt6ShaderToolsMacros.cmake:103 (_qt_internal_expose_source_file_to_ide)
  /usr/lib/cmake/Qt6ShaderToolsTools/Qt6ShaderToolsMacros.cmake:354 (_qt_internal_add_shaders_impl)
  src/rendergraph/shaders/CMakeLists.txt:17 (qt6_add_shaders)


rgba.vert -> rgba.vert.qsb exposed as :/shaders/rendergraph/rgba.vert.qsb
CMake Error at /usr/lib/cmake/Qt6Core/Qt6CoreMacros.cmake:2127 (get_target_property):
  get_target_property() called with non-existent target "rendergraph_sg".
Call Stack (most recent call first):
  /usr/lib/cmake/Qt6ShaderToolsTools/Qt6ShaderToolsMacros.cmake:103 (_qt_internal_expose_source_file_to_ide)
  /usr/lib/cmake/Qt6ShaderToolsTools/Qt6ShaderToolsMacros.cmake:354 (_qt_internal_add_shaders_impl)
  src/rendergraph/shaders/CMakeLists.txt:17 (qt6_add_shaders)


texture.frag -> texture.frag.qsb exposed as :/shaders/rendergraph/texture.frag.qsb
CMake Error at /usr/lib/cmake/Qt6Core/Qt6CoreMacros.cmake:2127 (get_target_property):
  get_target_property() called with non-existent target "rendergraph_sg".
Call Stack (most recent call first):
  /usr/lib/cmake/Qt6ShaderToolsTools/Qt6ShaderToolsMacros.cmake:103 (_qt_internal_expose_source_file_to_ide)
  /usr/lib/cmake/Qt6ShaderToolsTools/Qt6ShaderToolsMacros.cmake:354 (_qt_internal_add_shaders_impl)
  src/rendergraph/shaders/CMakeLists.txt:17 (qt6_add_shaders)


texture.vert -> texture.vert.qsb exposed as :/shaders/rendergraph/texture.vert.qsb
CMake Error at /usr/lib/cmake/Qt6Core/Qt6CoreMacros.cmake:2127 (get_target_property):
  get_target_property() called with non-existent target "rendergraph_sg".
Call Stack (most recent call first):
  /usr/lib/cmake/Qt6ShaderToolsTools/Qt6ShaderToolsMacros.cmake:103 (_qt_internal_expose_source_file_to_ide)
  /usr/lib/cmake/Qt6ShaderToolsTools/Qt6ShaderToolsMacros.cmake:354 (_qt_internal_add_shaders_impl)
  src/rendergraph/shaders/CMakeLists.txt:17 (qt6_add_shaders)


unicolor.frag -> unicolor.frag.qsb exposed as :/shaders/rendergraph/unicolor.frag.qsb
CMake Error at /usr/lib/cmake/Qt6Core/Qt6CoreMacros.cmake:2127 (get_target_property):
  get_target_property() called with non-existent target "rendergraph_sg".
Call Stack (most recent call first):
  /usr/lib/cmake/Qt6ShaderToolsTools/Qt6ShaderToolsMacros.cmake:103 (_qt_internal_expose_source_file_to_ide)
  /usr/lib/cmake/Qt6ShaderToolsTools/Qt6ShaderToolsMacros.cmake:354 (_qt_internal_add_shaders_impl)
  src/rendergraph/shaders/CMakeLists.txt:17 (qt6_add_shaders)


unicolor.vert -> unicolor.vert.qsb exposed as :/shaders/rendergraph/unicolor.vert.qsb
-- Adding qsb shaders to rendergraph_gl
pattern.frag -> pattern.frag.qsb exposed as :/shaders/rendergraph/pattern.frag.qsb
pattern.vert -> pattern.vert.qsb exposed as :/shaders/rendergraph/pattern.vert.qsb
rgb.frag -> rgb.frag.qsb exposed as :/shaders/rendergraph/rgb.frag.qsb
rgb.vert -> rgb.vert.qsb exposed as :/shaders/rendergraph/rgb.vert.qsb
rgba.frag -> rgba.frag.qsb exposed as :/shaders/rendergraph/rgba.frag.qsb
rgba.vert -> rgba.vert.qsb exposed as :/shaders/rendergraph/rgba.vert.qsb
texture.frag -> texture.frag.qsb exposed as :/shaders/rendergraph/texture.frag.qsb
texture.vert -> texture.vert.qsb exposed as :/shaders/rendergraph/texture.vert.qsb
unicolor.frag -> unicolor.frag.qsb exposed as :/shaders/rendergraph/unicolor.frag.qsb
unicolor.vert -> unicolor.vert.qsb exposed as :/shaders/rendergraph/unicolor.vert.qsb
-- Found wavpack: /usr/lib/libwavpack.so (found version "5.7.0")
-- Configuring incomplete, errors occurred!
```
</details>

Holzhaus commented on src/rendergraph/CMakeLists.txt:
```diff
@@ -39,5 +39,8 @@ set(
 
 add_subdirectory(opengl)
-add_subdirectory(scenegraph)
+#################################
+# TODO: uncomment in follow-up PR
+# add_subdirectory(scenegraph)
+#################################
```
Looks like this is the culprit that breaks it for me. WIthout `scenegraph`, the calls that reference the non-existing `rendergraph_sg` target fail.

Side note: Splitting the CMake code over multiple small `CMakeLists.txt` files that reference targets and variables defined elsewhere makes the control flow quite difficult to follow. Have you considered just using a single `CMakeLists.txt` inside the `rendergraph` directory?

m0dB replied:
Ok, for now the calls that reference the non-existing rendergraph_sg target fail should be commented out. I will do that.

I obsoletely prefer splitting cmake over multiple files, one per target.

m0dB replied:
Could you please try again, @Holzhaus ? I added conditionals to only add the shaders for the targets that were added.
(It looks like this is a change in Qt, being more strict about this in more recent versions)

Commit:
only build shaders for added target

Holzhaus reviewed (approved):
Build error is fixed now.

acolombier reviewed (approved):
Ran a smoke test and LGTM! Can't wait to have the rendergraph all out, thanks for your continuous efforts!

m0dB commented:
This are the follow up PRs that port the waveformrenderer code to use the rendergraph API:

stem waveformrenderer using rendergraph 
#14192 

slip waveformrenderer using rendergraph 
#14191

preroll waveformrenderer using rendergraph 
#14190

markrange waveformrenderer using rendergraph 
#14189

mark waveformrenderer using rendergraph 
#14188

endoftrack waveformrenderer using rendergraph 
#14187

signal waveformrenderer using rendergraph 
#14186

beat waveformrenderer using rendergraph 
#14185

When these have all been merged, we can also compile the scenegraph backend (rendergraph_sg) and @acolombier can use the above from QML.

acolombier commented:
Yes, this is - the shaders are packed inside the QRC

PPN-SD commented:
Hi,

Since PR is merged, I get errors during link time.

I need to set -DBUILD_SHARED_LIBS=OFF to fix it.

```
ld.lld: error: undefined symbol: vtable for rendergraph_gl::BaseNode
 referenced by waveformrendererstem.cpp
               waveformrendererstem.cpp.o:(allshader::WaveformRendererStem::WaveformRendererStem(WaveformWidgetRenderer*, WaveformRendererAbstract::PositionSour
ce)) in archive libmixxx-lib.a
 referenced by waveformrendermark.cpp
               waveformrendermark.cpp.o:(allshader::WaveformRenderMark::WaveformRenderMark(WaveformWidgetRenderer*, WaveformRendererAbstract::PositionSource)) i
n archive libmixxx-lib.a
 referenced by waveformrenderbackground.cpp
               waveformrenderbackground.cpp.o:(allshader::WaveformRenderBackground::WaveformRenderBackground(WaveformWidgetRenderer*)) in archive libmixxx-lib.a
 referenced 8 more times
 the vtable symbol may be undefined because the class is missing its key function (see https://lld.llvm.org/missingkeyfunction)

ld.lld: error: undefined symbol: vtable for rendergraph_gl::BaseOpenGLNode
 referenced by waveformrendererstem.cpp
               waveformrendererstem.cpp.o:(allshader::WaveformRendererStem::~WaveformRendererStem()) in archive libmixxx-lib.a
 referenced by waveformrendererstem.cpp
               waveformrendererstem.cpp.o:(allshader::WaveformRendererStem::WaveformRendererStem(WaveformWidgetRenderer*, WaveformRendererAbstract::PositionSource) (.cold)) in archive libmixxx-lib.a
 referenced by waveformrendermark.cpp
               waveformrendermark.cpp.o:(allshader::WaveformRenderMark::WaveformRenderMark(WaveformWidgetRenderer*, WaveformRendererAbstract::PositionSource) (.cold)) in archive libmixxx-lib.a
 referenced 47 more times
 the vtable symbol may be undefined because the class is missing its key function (see https://lld.llvm.org/missingkeyfunction)

ld.lld: error: undefined symbol: rendergraph_gl::BaseNode::~BaseNode()
 referenced by waveformrendererstem.cpp
               waveformrendererstem.cpp.o:(allshader::WaveformRendererStem::~WaveformRendererStem()) in archive libmixxx-lib.a
 referenced by waveformrendererstem.cpp
               waveformrendererstem.cpp.o:(allshader::WaveformRendererStem::WaveformRendererStem(WaveformWidgetRenderer*, WaveformRendererAbstract::PositionSource) (.cold)) in archive libmixxx-lib.a
 referenced by waveformrendererslipmode.cpp
               waveformrendererslipmode.cpp.o:(allshader::WaveformRendererSlipMode::~WaveformRendererSlipMode()) in archive libmixxx-lib.a
 referenced 66 more times

ld.lld: error: undefined symbol: typeinfo for rendergraph_gl::BaseOpenGLNode
 referenced by waveformrendererstem.cpp
               waveformrendererstem.cpp.o:(typeinfo for allshader::WaveformRendererStem) in archive libmixxx-lib.a
 referenced by waveformrenderersimple.cpp
               waveformrenderersimple.cpp.o:(typeinfo for allshader::WaveformRendererSimple) in archive libmixxx-lib.a
 referenced by waveformrendererrgb.cpp
               waveformrendererrgb.cpp.o:(typeinfo for allshader::WaveformRendererRGB) in archive libmixxx-lib.a
 referenced 8 more times
ld.lld: error: undefined symbol: rendergraph_gl::BaseOpenGLNode::render()
 referenced by waveformrendererstem.cpp
               waveformrendererstem.cpp.o:(vtable for allshader::WaveformRendererStem) in archive libmixxx-lib.a
 referenced by waveformrenderersimple.cpp
               waveformrenderersimple.cpp.o:(vtable for allshader::WaveformRendererSimple) in archive libmixxx-lib.a
 referenced by waveformrendererrgb.cpp
               waveformrendererrgb.cpp.o:(vtable for allshader::WaveformRendererRGB) in archive libmixxx-lib.a
 referenced 8 more times

ld.lld: error: undefined symbol: rendergraph_gl::BaseOpenGLNode::initialize()
 referenced by waveformrendererstem.cpp
               waveformrendererstem.cpp.o:(vtable for allshader::WaveformRendererStem) in archive libmixxx-lib.a
 referenced by waveformrenderersimple.cpp
               waveformrenderersimple.cpp.o:(vtable for allshader::WaveformRendererSimple) in archive libmixxx-lib.a
 referenced by waveformrendererrgb.cpp
               waveformrendererrgb.cpp.o:(vtable for allshader::WaveformRendererRGB) in archive libmixxx-lib.a
 referenced 8 more times

ld.lld: error: undefined symbol: rendergraph_gl::BaseOpenGLNode::resize(int, int)
 referenced by waveformrendererstem.cpp
               waveformrendererstem.cpp.o:(vtable for allshader::WaveformRendererStem) in archive libmixxx-lib.a
 referenced by waveformrenderersimple.cpp
               waveformrenderersimple.cpp.o:(vtable for allshader::WaveformRendererSimple) in archive libmixxx-lib.a
 referenced by waveformrendererrgb.cpp
               waveformrendererrgb.cpp.o:(vtable for allshader::WaveformRendererRGB) in archive libmixxx-lib.a
 referenced 8 more times

ld.lld: error: undefined symbol: rendergraph_gl::GeometryNode::GeometryNode()
 referenced by waveformrendererslipmode.cpp
               waveformrendererslipmode.cpp.o:(allshader::WaveformRendererSlipMode::WaveformRendererSlipMode(WaveformWidgetRenderer*)) in archive libmixxx-lib.a
 referenced by waveformrenderbeat.cpp
               waveformrenderbeat.cpp.o:(allshader::WaveformRenderBeat::WaveformRenderBeat(WaveformWidgetRenderer*, WaveformRendererAbstract::PositionSource)) in archive libmixxx-lib.a

ld.lld: error: undefined symbol: rendergraph_gl::GeometryNode::setUsePreprocess(bool)
 referenced by waveformrendererslipmode.cpp
               waveformrendererslipmode.cpp.o:(allshader::WaveformRendererSlipMode::WaveformRendererSlipMode(WaveformWidgetRenderer*)) in archive libmixxx-lib.a
 referenced by waveformrenderbeat.cpp
               waveformrenderbeat.cpp.o:(allshader::WaveformRenderBeat::WaveformRenderBeat(WaveformWidgetRenderer*, WaveformRendererAbstract::PositionSource)) in archive libmixxx-lib.a
```

PPN-SD commented:
@m0dB @acolombier As there are many changes waiting and coming on this feature, have you had time to look at these errors?

acolombier commented:
Sorry for missing the first notification - I had noticed similar issues when merging. I believe this was due to the namespace alias change; `allshader` shouldn't exist anymore as it is now a macros, you should have `allshader_gl` and `allshader_sg` instead (so `allshader_gl::WaveformRenderBeat::WaveformRenderBeat...` and `allshader_sg::WaveformRenderBeat::WaveformRenderBeat...`) Have you tried to clear your build directory and your ccache? If yes, could you please provide what version of GCC/Clang/MVC/? your are using?

PPN-SD commented:
Freshly test from git, clean env (VM), with GCC-14/15 without ccache : same errors.
You have noticed that these errors only occur with -DBUILD_SHARED_LIBS=ON.
